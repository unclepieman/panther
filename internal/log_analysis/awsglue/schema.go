package awsglue

/**
 * Panther is a Cloud-Native SIEM for the Modern Security Team.
 * Copyright (C) 2020 Panther Labs Inc
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

// Infers Glue table column types from Go types, recursively descends types

import (
	"github.com/panther-labs/panther/internal/log_analysis/awsglue/glueschema"
)

type Column = glueschema.Column

var (
	// RuleMatchColumns are columns added by the rules engine
	RuleMatchColumns = []Column{
		{
			Name:    "p_rule_id",
			Type:    glueschema.TypeString,
			Comment: "Rule id",
		},
		{
			Name:    "p_alert_id",
			Type:    glueschema.TypeString,
			Comment: "Alert id",
		},
		{
			Name:    "p_alert_context",
			Type:    glueschema.TypeString,
			Comment: "Additional alert context",
		},
		{
			Name:    "p_alert_creation_time",
			Type:    glueschema.TypeTimestamp,
			Comment: "The time the alert was initially created (first match)",
		},
		{
			Name:    "p_alert_update_time",
			Type:    glueschema.TypeTimestamp,
			Comment: "The time the alert last updated (last match)",
		},
		{
			Name:    "p_rule_tags",
			Type:    glueschema.ArrayOf(glueschema.TypeString),
			Comment: "The tags of the rule that generated this alert",
		},
		{
			Name:    "p_rule_reports",
			Type:    glueschema.MapOf(glueschema.TypeString, glueschema.ArrayOf(glueschema.TypeString)),
			Comment: "The reporting tags of the rule that generated this alert",
		},
	}

	// RuleErrorColumns are columns added by the rules engine
	RuleErrorColumns = append(
		RuleMatchColumns,
		Column{
			Name:    "p_rule_error",
			Type:    glueschema.TypeString,
			Comment: "The rule error",
		},
	)
)
