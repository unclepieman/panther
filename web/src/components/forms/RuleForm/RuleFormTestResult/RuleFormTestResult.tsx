/**
 * Panther is a Cloud-Native SIEM for the Modern Security Team.
 * Copyright (C) 2020 Panther Labs Inc
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import React from 'react';
import { ComplianceStatusEnum, TestRuleRecord } from 'Generated/schema';
import { Card, Flex, Box, Heading, Text, Grid } from 'pouncejs';
import { capitalize } from 'Helpers/utils';
import StatusBadge from 'Components/badges/StatusBadge';

interface RuleFormTestResultProps {
  testResult: TestRuleRecord;
}

const severityToString = (severity: string) => capitalize(severity.toLowerCase());

const RuleFormTestResult: React.FC<RuleFormTestResultProps> = ({ testResult }) => {
  const {
    functions: {
      ruleFunction,
      dedupFunction,
      titleFunction,
      alertContextFunction,
      referenceFunction,
      runbookFunction,
      descriptionFunction,
      severityFunction,
      destinationsFunction,
    },
    passed,
    name,
    error: unknownError,
  } = testResult;

  return (
    <Card p={4} as="article">
      <Flex align="flex-start" spacing={4}>
        <StatusBadge status={passed ? ComplianceStatusEnum.Pass : ComplianceStatusEnum.Fail} />
        <Box spacing={2}>
          <Heading as="h2" size="x-small" fontWeight="medium">
            {name}
          </Heading>
          <Grid
            as="dl"
            wordBreak="break-word"
            templateColumns="max-content 1fr"
            fontSize="medium"
            fontWeight="medium"
            columnGap={4}
            rowGap={2}
            mt={2}
          >
            {unknownError && (
              <React.Fragment>
                <Box as="dt" color="navyblue-100">
                  Error
                </Box>
                <Text as="dd" color="red-200">
                  {unknownError.message}
                </Text>
              </React.Fragment>
            )}
            {ruleFunction?.error && (
              <React.Fragment>
                <Box as="dt" color="navyblue-100">
                  Rule Body
                </Box>
                <Text as="dd" color="red-200">
                  {ruleFunction.error.message}
                </Text>
              </React.Fragment>
            )}
            {titleFunction && (
              <React.Fragment>
                <Box as="dt" color="navyblue-100">
                  Alert Title
                </Box>
                {!titleFunction.error ? (
                  <Text as="dd">{titleFunction.output}</Text>
                ) : (
                  <Text as="dd" color="red-200">
                    {titleFunction.error.message}
                  </Text>
                )}
              </React.Fragment>
            )}
            {dedupFunction && (
              <React.Fragment>
                <Box as="dt" color="navyblue-100">
                  Dedup String
                </Box>
                {!dedupFunction.error ? (
                  <Text as="dd">{dedupFunction.output}</Text>
                ) : (
                  <Text as="dd" color="red-200">
                    {dedupFunction.error.message}
                  </Text>
                )}
              </React.Fragment>
            )}

            {referenceFunction && (
              <React.Fragment>
                <Box as="dt" color="navyblue-100">
                  Reference
                </Box>
                {!referenceFunction.error ? (
                  <Text as="dd">{referenceFunction.output}</Text>
                ) : (
                  <Text as="dd" color="red-200">
                    {referenceFunction.error.message}
                  </Text>
                )}
              </React.Fragment>
            )}
            {runbookFunction && (
              <React.Fragment>
                <Box as="dt" color="navyblue-100">
                  Runbook
                </Box>
                {!runbookFunction.error ? (
                  <Text as="dd">{runbookFunction.output}</Text>
                ) : (
                  <Text as="dd" color="red-200">
                    {runbookFunction.error.message}
                  </Text>
                )}
              </React.Fragment>
            )}
            {descriptionFunction && (
              <React.Fragment>
                <Box as="dt" color="navyblue-100">
                  Description
                </Box>
                {!descriptionFunction.error ? (
                  <Text as="dd">{descriptionFunction.output}</Text>
                ) : (
                  <Text as="dd" color="red-200">
                    {descriptionFunction.error.message}
                  </Text>
                )}
              </React.Fragment>
            )}
            {severityFunction && (
              <React.Fragment>
                <Box as="dt" color="navyblue-100">
                  Severity
                </Box>
                {!severityFunction.error ? (
                  <Text as="dd">{severityToString(severityFunction.output)}</Text>
                ) : (
                  <Text as="dd" color="red-200">
                    {severityFunction.error.message}
                  </Text>
                )}
              </React.Fragment>
            )}
            {destinationsFunction && (
              <React.Fragment>
                <Box as="dt" color="navyblue-100">
                  Destination Overrides
                </Box>
                {!destinationsFunction.error ? (
                  <Text as="dd">{destinationsFunction.output}</Text>
                ) : (
                  <Text as="dd" color="red-200">
                    {destinationsFunction.error.message}
                  </Text>
                )}
              </React.Fragment>
            )}
            {alertContextFunction && (
              <React.Fragment>
                <Box as="dt" color="navyblue-100">
                  Alert Context
                </Box>
                {!alertContextFunction.error ? (
                  <Text as="dd">{alertContextFunction.output}</Text>
                ) : (
                  <Text as="dd" color="red-200">
                    {alertContextFunction.error.message}
                  </Text>
                )}
              </React.Fragment>
            )}
          </Grid>
        </Box>
      </Flex>
    </Card>
  );
};

export default RuleFormTestResult;
