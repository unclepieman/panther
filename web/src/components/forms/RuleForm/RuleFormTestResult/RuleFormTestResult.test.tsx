/**
 * Panther is a Cloud-Native SIEM for the Modern Security Team.
 * Copyright (C) 2020 Panther Labs Inc
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import React from 'react';
import {
  buildError,
  buildTestRuleRecord,
  buildTestRuleRecordFunctions,
  buildTestDetectionSubRecord,
  render,
} from 'test-utils';
import { ComplianceStatusEnum, SeverityEnum } from 'Generated/schema';
import RuleFormTestResult from './RuleFormTestResult';

describe('RuleFormTestResult', () => {
  it('shows the name & status of the test', () => {
    const testResult = buildTestRuleRecord({ passed: true });

    const { getByText } = render(<RuleFormTestResult testResult={testResult} />);
    expect(getByText(testResult.name)).toBeInTheDocument();
    expect(getByText(ComplianceStatusEnum.Pass)).toBeInTheDocument();
  });

  it('shows a generic error when it exists', () => {
    const testResult = buildTestRuleRecord({
      functions: {
        ruleFunction: null,
        titleFunction: null,
        dedupFunction: null,
      },
    });

    const { getByText } = render(<RuleFormTestResult testResult={testResult} />);
    expect(getByText(testResult.error.message)).toBeInTheDocument();
  });

  it('shows a list of all the non-generic errors', () => {
    const testResult = buildTestRuleRecord({
      error: null,
      functions: buildTestRuleRecordFunctions({
        ruleFunction: buildTestDetectionSubRecord({ error: buildError({ message: 'Rule' }) }),
        titleFunction: buildTestDetectionSubRecord({ error: buildError({ message: 'Title' }) }),
        dedupFunction: buildTestDetectionSubRecord({ error: buildError({ message: 'Dedup' }) }),
      }),
    });

    const { getByText } = render(<RuleFormTestResult testResult={testResult} />);
    expect(getByText(testResult.functions.ruleFunction.error.message)).toBeInTheDocument();
    expect(getByText(testResult.functions.titleFunction.error.message)).toBeInTheDocument();
    expect(getByText(testResult.functions.dedupFunction.error.message)).toBeInTheDocument();
  });

  it('shows a list of all the non-generic errors for all function outputs', () => {
    const testResult = buildTestRuleRecord({
      error: null,
      functions: buildTestRuleRecordFunctions({
        ruleFunction: buildTestDetectionSubRecord({ error: buildError({ message: 'Rule' }) }),
        titleFunction: buildTestDetectionSubRecord({ error: buildError({ message: 'Title' }) }),
        dedupFunction: buildTestDetectionSubRecord({
          error: buildError({ message: 'Dedup Error' }),
        }),
        referenceFunction: buildTestDetectionSubRecord({
          error: buildError({ message: 'Ref Error' }),
        }),
        runbookFunction: buildTestDetectionSubRecord({
          error: buildError({ message: 'Runbook Error' }),
        }),
        destinationsFunction: buildTestDetectionSubRecord({
          error: buildError({ message: 'Dest Error' }),
        }),
        descriptionFunction: buildTestDetectionSubRecord({
          error: buildError({ message: 'Desc Error' }),
        }),
        severityFunction: buildTestDetectionSubRecord({
          error: buildError({ message: 'Severity Error' }),
        }),
      }),
    });

    const { getByText } = render(<RuleFormTestResult testResult={testResult} />);
    expect(getByText(testResult.functions.ruleFunction.error.message)).toBeInTheDocument();
    expect(getByText(testResult.functions.titleFunction.error.message)).toBeInTheDocument();
    expect(getByText(testResult.functions.dedupFunction.error.message)).toBeInTheDocument();
    expect(getByText(testResult.functions.referenceFunction.error.message)).toBeInTheDocument();
    expect(getByText(testResult.functions.runbookFunction.error.message)).toBeInTheDocument();
    expect(getByText(testResult.functions.destinationsFunction.error.message)).toBeInTheDocument();
    expect(getByText(testResult.functions.descriptionFunction.error.message)).toBeInTheDocument();
    expect(getByText(testResult.functions.severityFunction.error.message)).toBeInTheDocument();
  });

  it("shows functional outputs when errors don't exist", () => {
    const testResult = buildTestRuleRecord({
      error: null,
      functions: buildTestRuleRecordFunctions({
        titleFunction: buildTestDetectionSubRecord({ output: 'Title', error: null }),
        dedupFunction: buildTestDetectionSubRecord({ output: 'Dedup', error: null }),
        referenceFunction: buildTestDetectionSubRecord({ output: 'Reference output', error: null }),
        runbookFunction: buildTestDetectionSubRecord({ output: 'This is a runbook', error: null }),
        descriptionFunction: buildTestDetectionSubRecord({
          output: 'This is a really big description with lots of metadata',
          error: null,
        }),
        // Severity can be one of SeverityEnum but we capitalize the output
        severityFunction: buildTestDetectionSubRecord({
          output: SeverityEnum.Low,
          error: null,
        }),
        destinationsFunction: buildTestDetectionSubRecord({
          output: 'Slack Destination',
          error: null,
        }),
      }),
    });

    const { getByText } = render(<RuleFormTestResult testResult={testResult} />);
    expect(getByText(testResult.functions.titleFunction.output)).toBeInTheDocument();
    expect(getByText(testResult.functions.dedupFunction.output)).toBeInTheDocument();
    expect(getByText(testResult.functions.runbookFunction.output)).toBeInTheDocument();
    expect(getByText(testResult.functions.descriptionFunction.output)).toBeInTheDocument();
    expect(getByText('Low')).toBeInTheDocument();
    expect(getByText(testResult.functions.referenceFunction.output)).toBeInTheDocument();
    expect(getByText(testResult.functions.destinationsFunction.output)).toBeInTheDocument();
  });

  it('matches the snapshot', () => {
    const testResult = buildTestRuleRecord();

    const { container } = render(<RuleFormTestResult testResult={testResult} />);
    expect(container).toMatchSnapshot();
  });
});
