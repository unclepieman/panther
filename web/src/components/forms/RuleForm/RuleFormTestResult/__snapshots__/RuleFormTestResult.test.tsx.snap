// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RuleFormTestResult matches the snapshot 1`] = `
.panther-25 {
  background-color: #23344e;
  border-radius: 4px;
  padding: 16px;
}

.panther-24 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

.panther-24>*:not(:last-child) {
  margin-right: 16px;
}

.panther-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  cursor: default;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  min-width: 85px;
  text-align: center;
  font-weight: 700;
  border-radius: 40px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  font-size: 0.75rem;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 16px;
  padding-right: 16px;
  background-color: #50931d;
}



.panther-1 {
  font-weight: 500;
  font-size: 1.125rem;
}

.panther-22 {
  display: grid;
  grid-row-gap: 8px;
  grid-column-gap: 16px;
  grid-template-columns: max-content 1fr;
  word-break: break-word;
  font-size: 0.875rem;
  font-weight: 500;
  margin-top: 8px;
}

.panther-2 {
  color: #86a3c3;
}

.panther-3 {
  color: #e85959;
}

<div>
  <header
    id="main-header"
  />
  <article
    class="panther-25"
  >
    <div
      class="panther-24"
    >
      <div
        aria-atomic="true"
        class="panther-0"
        role="status"
      >
        PASS
      </div>
      <div
        class="panther-23"
        spacing="2"
      >
        <h2
          class="panther-1"
        >
          Pants
        </h2>
        <dl
          class="panther-22"
        >
          <dt
            class="panther-2"
          >
            Error
          </dt>
          <dd
            class="panther-3"
          >
            deposit
          </dd>
          <dt
            class="panther-2"
          >
            Rule Body
          </dt>
          <dd
            class="panther-3"
          >
            deposit
          </dd>
          <dt
            class="panther-2"
          >
            Alert Title
          </dt>
          <dd
            class="panther-3"
          >
            deposit
          </dd>
          <dt
            class="panther-2"
          >
            Dedup String
          </dt>
          <dd
            class="panther-3"
          >
            deposit
          </dd>
          <dt
            class="panther-2"
          >
            Reference
          </dt>
          <dd
            class="panther-3"
          >
            deposit
          </dd>
          <dt
            class="panther-2"
          >
            Runbook
          </dt>
          <dd
            class="panther-3"
          >
            deposit
          </dd>
          <dt
            class="panther-2"
          >
            Description
          </dt>
          <dd
            class="panther-3"
          >
            deposit
          </dd>
          <dt
            class="panther-2"
          >
            Severity
          </dt>
          <dd
            class="panther-3"
          >
            deposit
          </dd>
          <dt
            class="panther-2"
          >
            Destination Overrides
          </dt>
          <dd
            class="panther-3"
          >
            deposit
          </dd>
          <dt
            class="panther-2"
          >
            Alert Context
          </dt>
          <dd
            class="panther-3"
          >
            deposit
          </dd>
        </dl>
      </div>
    </div>
  </article>
  <footer
    id="footer"
  />
</div>
`;
